geomesa-index-api/src/main/scala/org/locationtech/geomesa/index/geotools/GeoMesaFeatureCollection.scala

```diff
@@ -11,96 +11,248 @@ package org.locationtech.geomesa.index.geotools
-    if (!open.get) {
-      // once opened the query will already be configured by the query planner,
-      // otherwise do it here
-      source.runner.configureQuery(source.getSchema, query)
+  override def sort(order: SortBy): SimpleFeatureCollection = {
+    val merged = new Query(original)
+    merged.setHints(new Hints(original.getHints))
+    if (merged.getSortBy == null) {
+      merged.setSortBy(Array(order))
+    } else {
+      merged.setSortBy(merged.getSortBy :+ order)
     }
-    source.getCount(query)
+    new GeoMesaFeatureCollection(source, merged)
   }
 
+  override def getBounds: ReferencedEnvelope = source.getBounds(query)
+
+  override def getCount: Int = source.getCount(configured)
+
   // note: this shouldn't return -1 (as opposed to FeatureSource.getCount), but we still don't return a valid
   // size unless exact counts are enabled
   override def size: Int = {
     val count = getCount
     if (count < 0) { 0 } else { count }
   }
+
+  override def contains(o: Any): Boolean = {
+    o match {
+      case f: SimpleFeature =>
+        val sub = subCollection(FilterHelper.ff.id(f.getIdentifier))
+        WithClose(CloseableIterator(sub.features()))(_.nonEmpty)
+
+      case _ => false
+    }
+  }
+
+  override def containsAll(collection: java.util.Collection[_]): Boolean = {
+    val size = collection.size()
+    if (size == 0) { true } else {
+      val filters = Seq.newBuilder[Filter]
+      filters.sizeHint(size)
+
+      val features = collection.iterator()
+      while (features.hasNext) {
+        features.next match {
+          case f: SimpleFeature => filters += FilterHelper.ff.id(f.getIdentifier)
+          case _ => return false
+        }
+      }
+
+      val sub = subCollection(org.locationtech.geomesa.filter.orFilters(filters.result))
+      WithClose(CloseableIterator(sub.features()))(_.length) == size
+    }
+  }
 }
 
-object GeoMesaFeatureCollection {
+object GeoMesaFeatureCollection extends LazyLogging {
+
   private val oneUp = new AtomicLong(0)
+
   def nextId: String = s"GeoMesaFeatureCollection-${oneUp.getAndIncrement()}"
+
+  /**
+    * Attempts to visit the feature collection in an optimized manner. This will unwrap any decorating
+    * feature collections that may interfere with the `accepts` method.
+    *
+    * Note that generally this may not be a good idea - collections are presumably wrapped for a reason.
+    * However, our visitation functionality keeps being broken by changes in GeoServer, so we're being
+    * defensive here.
+    *
+    * @param collection feature collection
+    * @param visitor visitor
+    * @param progress progress monitor
+    */
+  def visit(
+      collection: FeatureCollection[SimpleFeatureType, SimpleFeature],
+      visitor: FeatureVisitor,
+      progress: ProgressListener = new NullProgressListener): Unit = {
+    val unwrapped = if (collection.isInstanceOf[GeoMesaFeatureVisitingCollection]) { collection } else {
+      try { unwrap(collection) } catch {
```

Comment:
Is it worth adding another logging statement or two here?  I suppose I'm suggesting one for the case where we see a GeoMesaFeatureVisitingCollection and do nothing.

Reply:
I believe all our processes log the feature collection class, so it would be somewhat redundant, what do you think?

Reply:
That makes sense.  I'm being overly defensive;).